ompi/mca/osc/ucx/osc_ucx_component.c

```diff
@@ -0,0 +1,565 @@
+                                        module->win_info_array, flavor);
+    if (ret != OMPI_SUCCESS) {
+        goto error;
+    }
+
+    /* initialize UCP state_info array */
+    module->state_info_array = calloc(comm_size, sizeof(ompi_osc_ucx_win_info_t));
+    if (module->state_info_array == NULL) {
+        goto error;
+    }
+
+    state_base = (void *)&(module->state);
+    ret = mem_map_and_exchange_win_info(&state_base, sizeof(ompi_osc_ucx_state_t),
+                                        &(module->state_memh), module,
+                                        module->state_info_array,
+                                        MPI_WIN_FLAVOR_CREATE);
+    if (ret != OMPI_SUCCESS) {
+        goto error;
+    }
+
+    module->state.lock = TARGET_LOCK_UNLOCKED;
+    module->state.post_count = 0;
+    module->state.complete_count = 0;
+    module->epoch_type.access = NONE_EPOCH;
+    module->epoch_type.exposure = NONE_EPOCH;
+    module->lock_count = 0;
+    module->start_group = NULL;
+    module->post_group = NULL;
+    OBJ_CONSTRUCT(&module->outstanding_locks, opal_list_t);
+    module->global_ops_num = 0;
+    module->per_target_ops_nums = calloc(comm_size, sizeof(int));
+
+    win->w_osc_module = &module->super;
+
+    /* sync with everyone */
+
+    ret = module->comm->c_coll->coll_barrier(module->comm,
+                                             module->comm->c_coll->coll_barrier_module);
+
+    if (ret != OMPI_SUCCESS) {
+        goto error;
+    }
+
+    return ret;
+
+ error:
+    if (my_addr) ucp_worker_release_address(mca_osc_ucx_component.ucp_worker, my_addr);
+    if (recv_buf) free(recv_buf);
+    if (module->win_info_array) free(module->win_info_array);
+    if (module->state_info_array) free(module->state_info_array);
+    if (module->disp_units) free(module->disp_units);
+    if (module->comm) ompi_comm_free(&module->comm);
+    if (module->per_target_ops_nums) free(module->per_target_ops_nums);
+    if (module) free(module);
+    return ret;
+}
+
+int ompi_osc_ucx_win_attach(struct ompi_win_t *win, void *base, size_t len) {
+    return OMPI_SUCCESS;
+}
+
+int ompi_osc_ucx_win_detach(struct ompi_win_t *win, const void *base) {
+    return OMPI_SUCCESS;
+}
+
+int ompi_osc_ucx_free(struct ompi_win_t *win) {
+    ompi_osc_ucx_module_t *module = (ompi_osc_ucx_module_t*) win->w_osc_module;
+    int i, ret = OMPI_SUCCESS;
+
+    if ((module->epoch_type.access != NONE_EPOCH &&
+         module->epoch_type.access != FENCE_EPOCH) ||
+        module->epoch_type.exposure != NONE_EPOCH) {
+        ret = OMPI_ERR_RMA_SYNC;
+    }
+
+    if (module->start_group != NULL || module->post_group != NULL) {
+        ret = OMPI_ERR_RMA_SYNC;
+    }
+
+    assert(module->global_ops_num == 0);
+    assert(module->lock_count == 0);
+    assert(opal_list_is_empty(&module->outstanding_locks) == true);
+    OBJ_DESTRUCT(&module->outstanding_locks);
+
+    while (module->state.lock != TARGET_LOCK_UNLOCKED) {
```

Comment:
(now L669)
I think it is not an error. For example: 
rank0-----------rank1

win_lock
put/get
win_unlock
win_free-------win_free

"module->state.lock" is the lock state on the target side (rank1). Rank1 should wait until this state changed to be unlocked, which means rank0 finished all osc, then free the window.